File: app/src/main/java/d/d/meshenger/PacketWriter.java
package d.d.meshenger;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;


/* Write the message header before the message is send */
class PacketWriter {
    final OutputStream os;
    final byte[] header;

    public PacketWriter(Socket socket) throws IOException {
        this.os = socket.getOutputStream();
        this.header = new byte[4];
    }

    private static void writeMessageHeader(byte[] packet, int value) {
        packet[0] = (byte) ((value >> 24) & 0xff);
        packet[1] = (byte) ((value >> 16) & 0xff);
        packet[2] = (byte) ((value >> 8) & 0xff);
        packet[3] = (byte) ((value >> 0) & 0xff);
    }

    public void writeMessage(byte[] message) throws IOException {
    	writeMessageHeader(this.header, message.length);
    	// need to concatenate?
    	this.os.write(this.header);
        this.os.write(message);
    }

    private void log(String s) {
        Log.d(this, s);
    }
}

File: app/src/main/java/d/d/meshenger/PacketWriter.kt
package d.d.meshenger

import java.io.IOException
import java.net.Socket

/* Write the message header before the message is send */
internal class PacketWriter(socket: Socket) {
    val os = socket.getOutputStream()
    val header = ByteArray(4)

    @Throws(IOException::class)
    fun writeMessage(message: ByteArray) {
        writeMessageHeader(header, message.size)
        // need to concatenate?
        os.write(header)
        os.write(message)
    }

    companion object {
        private fun writeMessageHeader(packet: ByteArray, value: Int) {
            packet[0] = (value shr 24 and 0xff).toByte()
            packet[1] = (value shr 16 and 0xff).toByte()
            packet[2] = (value shr 8 and 0xff).toByte()
            packet[3] = (value shr 0 and 0xff).toByte()
        }
    }
}